lib/nfc/protocols/felica/felica_listener.c

```diff
@@ -49,6 +57,102 @@ const FelicaData* felica_listener_get_data(const FelicaListener* instance) {
     return instance->data;
 }
 
+static FelicaError felica_listener_command_handler_read(
+    FelicaListener* instance,
+    const FelicaListenerGenericRequest* const generic_request) {
+    const FelicaListenerReadRequest* request = (FelicaListenerReadRequest*)generic_request;
+    FURI_LOG_D(TAG, "Read cmd");
+
+    FelicaListenerReadCommandResponse* resp = malloc(
+        sizeof(FelicaCommandResponseHeader) + 1 +
+        FELICA_LISTENER_READ_BLOCK_COUNT_MAX * FELICA_DATA_BLOCK_SIZE);
+    furi_check(resp);
+
+    resp->header.response_code = FELICA_LISTENER_RESPONSE_CODE_READ;
+    resp->header.idm = request->base.header.idm;
+    resp->header.length = sizeof(FelicaCommandResponseHeader);
+
+    if(felica_listener_validate_read_request_and_set_sf(instance, request, &resp->header)) {
+        resp->block_count = request->base.header.block_count;
+        resp->header.length++;
+    } else {
+        resp->block_count = 0;
+    }
+
+    instance->mac_calc_start = 0;
+    memset(instance->requested_blocks, 0, sizeof(instance->requested_blocks));
+    const FelicaBlockListElement* item =
+        felica_listener_block_list_item_get_first(instance, request);
+    for(uint8_t i = 0; i < resp->block_count; i++) {
+        instance->requested_blocks[i] = item->block_number;
+        FelicaCommanReadBlockHandler handler =
+            felica_listener_get_read_block_handler(item->block_number);
+
+        handler(instance, item->block_number, i, resp);
+
+        item = felica_listener_block_list_item_get_next(instance, item);
+    }
+
+    bit_buffer_reset(instance->tx_buffer);
+    bit_buffer_append_bytes(instance->tx_buffer, (uint8_t*)resp, resp->header.length);
+    free(resp);
+
+    return felica_listener_frame_exchange(instance, instance->tx_buffer);
+}
+
+static FelicaError felica_listener_command_handler_write(
+    FelicaListener* instance,
+    const FelicaListenerGenericRequest* const generic_request) {
+    FURI_LOG_D(TAG, "Write cmd");
+
+    const FelicaListenerWriteRequest* request = (FelicaListenerWriteRequest*)generic_request;
+    const FelicaListenerWriteBlockData* data_ptr =
+        felica_listener_get_write_request_data_pointer(instance, generic_request);
+
+    FelicaListenerWriteCommandResponse* resp =
+        (FelicaListenerWriteCommandResponse*)malloc(sizeof(FelicaListenerWriteCommandResponse));
+    furi_check(resp);
```

Comment:
no need in furi_check(). Malloc will crash if there is no memory
Also unnecessary type cast